From 9644efd7de116f3528079daf5618d09ab5ad7b77 Mon Sep 17 00:00:00 2001
From: Johny Lin <johnylin@google.com>
Date: Thu, 11 Apr 2024 14:33:56 +0000
Subject: [PATCH] CHROMIUM: sof: mediatek: skip readback binary kcontrol

The getter function for binary kcontrols of SOF widgets may be a
no-op, i.e. returning 0 without a touch of read buffer. In such
circumstances the host will receive the unwanted data and taint the
kcontrol cache which the host relies on for DSP recovery across
suspend/resume.

Skip readback given that is unnecessary to all kcontrols on DSP.
Revert this commit when no no-op kcontrol is on DSP.

BUG=b:331302167
TEST=stress audio playback via Speaker and suspend/resume

Change-Id: I12c4f96efb9160803280928cc2364f465e311123
Signed-off-by: Johny Lin <johnylin@google.com>
---
 sound/soc/sof/ipc3-control.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/sound/soc/sof/ipc3-control.c b/sound/soc/sof/ipc3-control.c
index a8deec7dc021..99ed561f7ebb 100644
--- a/sound/soc/sof/ipc3-control.c
+++ b/sound/soc/sof/ipc3-control.c
@@ -686,11 +686,25 @@ static int sof_ipc3_widget_kcontrol_setup(struct snd_sof_dev *sdev,
 			if (swidget->dynamic_pipeline_widget)
 				continue;
 
+#ifdef CONFIG_SND_SOC_SOF_MTK_TOPLEVEL
+			/*
+			 * The getter function for binary kcontrols of SOF widgets may
+			 * be a no-op, i.e. returning 0 without a touch of read buffer.
+			 * In such circumstances the host will receive the unwanted data
+			 * and taint the kcontrol cache which the host relies on for DSP
+			 * recovery across suspend/resume.
+			 * TODO(b/331302167): Skip readback given that is unnecessary to
+			 *     all kcontrols on DSP. Revert this when no no-op kcontrol
+			 *     is on DSP.
+			 */
+			dev_info(sdev->dev, "skip readback kcontrol data anyway\n");
+#else
 			ret = sof_ipc3_set_get_kcontrol_data(scontrol, false, false);
 			if (ret < 0)
 				dev_warn(sdev->dev,
 					 "kcontrol %d read failed for widget %s\n",
 					 scontrol->comp_id, swidget->widget->name);
+#endif
 		}
 
 	return 0;
-- 
2.44.0.683.g7961c838ac-goog

